_posts/2016-04-26-distributed-game-of-life-in-elixir.md

```diff
@@ -0,0 +1,1065 @@
+                      GameOfLife.Board, :keep_alive_tick, [alive_cells])
+    become_alive_task = Task.Supervisor.async(
+                        {GameOfLife.TaskSupervisor, GameOfLife.NodeManager.random_node},
+                        GameOfLife.Board, :become_alive_tick, [alive_cells])
+
+    keep_alive_cells = Task.await(keep_alive_task)
+    born_cells = Task.await(become_alive_task)
+
+    alive_cells = keep_alive_cells ++ born_cells
+
+    {:noreply, {alive_cells, tref, generation_counter + 1}}
+  end
+{% endhighlight %}
+
+What we are doing here is spinning off a new async process to run `GameOfLife.keep_alive_tick/1` function with argument `alive_cells`.
+
+{% highlight elixir %}
+# lib/game_of_life/board_server.ex
+keep_alive_task = Task.Supervisor.async(
+                  {GameOfLife.TaskSupervisor, GameOfLife.NodeManager.random_node},
+                  GameOfLife.Board, :keep_alive_tick, [alive_cells])
+{% endhighlight %}
+
+The tuple `{GameOfLife.TaskSupervisor, GameOfLife.NodeManager.random_node}` tells `Task.Supervisor` that we want to use task supervisor with the name `GameOfLife.TaskSupervisor` and we want to run the process on the node returned by `GameOfLife.NodeManager.random_node` function.
+
+# Create node manager for task supervisor
+
+Let's start with something simple just to see if we can distribute work across nodes in the cluster. We assume each new process created by task supervisor will be assigned randomly to one of the connected nodes. Each node should be equally overloaded with the assumption that each task is pretty similar and all nodes are machines with the same configuration and overload.
+
+{% highlight elixir %}
+# lib/game_of_life/node_manager.ex
+defmodule GameOfLife.NodeManager do
+  def all_nodes do
+    [Node.self | Node.list]
+  end
+
+  def random_node do
+    all_nodes |> Enum.random
+  end
+end
+{% endhighlight %}
+
+Our node manager has `random_node/0` function which returns the name of a random node connected to the cluster. Basically, that's it. Simple solution should be enough for now.
+
+# Create board helper functions
+
+We need some helper functions for operations we can do on board like adding, removing cells. Let's start with tests for module `GameOfLife.Board` and function `add_cells/2`.
+
+{% highlight elixir %}
+# test/game_of_life/board_test.exs
+defmodule GameOfLife.BoardTest do
+  use ExUnit.Case, async: true
+
+  test "add new cells to alive cells without duplicates" do
+    alive_cells = [{1, 1}, {2, 2}]
+    new_cells = [{0, 0}, {1, 1}]
+    actual_alive_cells = GameOfLife.Board.add_cells(alive_cells, new_cells)
+                          |> Enum.sort
+    expected_alive_cells = [{0, 0}, {1, 1}, {2, 2}]
+    assert actual_alive_cells == expected_alive_cells
+  end
+end
+{% endhighlight %}
+
+We need to ensure we won't allow to add the same cell twice to the board hence the above test. Here is implementation for `add_cells/2` function:
```

Comment:
Maybe this will be better?

We need to ensure we won’t allow adding the same cell twice to the board hence we test there are no duplicates.